### ISSUE TITLE
zip action - support for include/exclude

### ISSUE THREAD

**Issue by pritidesai** [apache/incubator-openwhisk-wskdeploy#969]
In case of multiple actions with shared files like a library or utility, `wskdeploy` should provide a way to specify `include` or `exclude` for an action so that it can support creating zip actions with an action file plus common shared file without user duplicating that utility file in each action.

For example, add support for something like this:

```
ls -1 actions/common/
utility.js
ls -1 actions/action1/
action1.js
ls -1 actions/action2/
action2.js
ls -1 actions/action3/
action3.js
```

manifest.yaml

```
packages:
    myPackage:
        actions:
             action1: actions/action1
             runtime: nodejs:6
              include:
                  - common/utility.js
             action2: actions/action2
             runtime: nodejs:6
              include:
                  - common/utility.js
             action3: actions/action3
             runtime: nodejs:6
              include:
                  - common/utility.js
```

**Comment by pritidesai** [comment]
I found a gap in the above example where `common/utility.js` is under `actions/` so when its specified as is under `include` section, `common/utility.js` can not be found. Here is the directory structure with `manifest.yaml`:

```
ls -1
manifest.yaml
actions
ls -1 actions/common/
utility.js
ls -1 actions/action1/
action1.js
package.json
ls -1 actions/action2/
action2.js
package.json
ls -1 actions/action3/
action3.js
package.json
```

Also, zipped actions must contain either package.json or index.js at the root which means `actions/common/utility.js` has to be moved under `actions/action1`. Now, either we move *all* files listed under `include` under root dir specified at the function level or we add a `source` and `destination` terminology in the manifest.

**Comment by pritidesai** [comment]
Also, can not work for a directory structure like this:

```
ls -1 ./
actions/
manifest.yaml

ls -1 actions/
common/
action1.js
action2.js
action3.js
action4.js
package.json

ls -1 actions/common
utility.js
```

manifest.yaml:

```
packages:
    myPackage:
        actions:
             action1:
                 function: actions/
                 runtime: nodejs:6
                 exclude:
                     - actions/**
                 include:
                     - actions/common/utility.js
                     - actions/action1.js
                     - actions/package.json
             action2:
                 function: actions/
                 runtime: nodejs:6
                 exclude:
                     - actions/**
                 include:
                     - actions/common/utility.js
                     - actions/action2.js
                     - actions/package.json
...
```


There has to be separate `package.json` for each action.

**Comment by pritidesai** [comment]
Based on an example [here](https://forum.serverless.com/t/all-functions-included-in-package-but-needed-only-one/3612/2), serverless has `include/exclude` support at the package level.

**Comment by mrutkows** [comment]
This perhaps may be the way to go... (implicit source/destination treating each include line as a series of arguments/parameters into an "include" function):
# issues:
# copy file to path ending in '/' (or not ending in '/')
# do we allow rename? or allow only target/dest ONLY to be a Path (not a new filename)
# Warning on changing .js to .py? user resp. for making sure this does not happen
# wildcards allowed 
# support recursive copy? or only directory by directory (future functionality, new parameter?)
packages:
    myPackage:
        actions:
            action1: 
                function: actions/action1
                runtime: nodejs:6
                include:
                    - ["common/utility1.js", "./utils/"]
                    - ["common/utility2.js", "./utils"]
                    - ["common/*.js", "./utils"]
                    - ["common/utilityX.js", "./utils/utilZ.js"]

**Comment by mrutkows** [comment]
the unordered list will not support guarantied ordering, still need better syntax...

**Comment by Juice10** [comment]
Here is a sample ruby project use case from the [OpenWhisk Ruby runtime launch blogpost](https://medium.com/openwhisk/ruby-goes-serverless-apache-openwhisk-adds-native-support-for-ruby-7134faee14e#5d97).

```bash
$ ls .** **
.bundle/config
.git/HEAD
.git/config
.git/description
.git/hooks/applypatch-msg.sample
.git/hooks/commit-msg.sample
.git/hooks/fsmonitor-watchman.sample
.git/hooks/post-update.sample
.git/hooks/pre-applypatch.sample
.git/hooks/pre-commit.sample
.git/hooks/pre-push.sample
.git/hooks/pre-rebase.sample
.git/hooks/pre-receive.sample
.git/hooks/prepare-commit-msg.sample
.git/hooks/update.sample
.git/info/exclude
Gemfile
Gemfile.lock
bundle/bundler/setup.rb
bundle/ruby/2.5.0/cache/betterlorem-0.1.2.gem
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/doc/BetterLorem.html
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/doc/created.rid
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/doc/index.html
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/doc/lorem_txt.html
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/doc/rdoc.css
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/doc/table_of_contents.html
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/lib/betterlorem.rb
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/lib/lorem.txt
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/spec/betterlorem_spec.rb
bundle/ruby/2.5.0/gems/betterlorem-0.1.2/spec/spec_helper.rb
bundle/ruby/2.5.0/specifications/betterlorem-0.1.2.gemspec
main.rb
```

I'd love to see how we can represent this. 

-  First off we don't need all the .git stuff, we won't need the documentation or specs for the gems.
- Secondly we need to specify that `main.rb` is where our action is going to be executed. 
- As a sidenote, [these files](https://github.com/github/gitignore/blob/master/Ruby.gitignore) will probably often need excluding.

I looked at [package.json's files convention](https://docs.npmjs.com/files/package.json#files) for some inspiration and came up with this representation:

```
packages:
    myPackage:
        actions:
            lipsum: 
                path: .
                runtime: ruby:2.5
                main: main.rb
                exclude:
                    - .git/*
                    - bundle/ruby/*/gems/*/spec/
                    - bundle/**/docs/
```

**Comment by pritidesai** [comment]
thanks a bunch @Juice10 for inputs, good to have more use/test cases, this is the exact same syntax I am planning on supporting but let me confirm. Will have a PR soon.

**Comment by pritidesai** [comment]
Fixed with PR #991
